Storage: Add tax lines to Orders

selanthiraiyan opened the pull request:
Part of #5809

### Description
Second PR for the task of removing on-device tax calculation and displaying tax breakup details in Simple payments.

This PR adds support for `tax_lines` in Storage. There will be a follow up pull request raised to add support for `tax_lines` in Yosemite.

#### Changes
- Increment core-data model version and add `OrderTaxLine` entity.
- Configure relationships with `Order` and `OrderItemAttribute`.
- Add `taxes` property in `Order` to hold `OrderTaxLine`.
- Update `MIGRATIONS.md` about adding Model 61 for OrderTaxLine.
- Add `loadOrderTaxLine` to `StorageType` to load OrderTaxLine.
- Add test to validate `loadOrderTaxLine` method.
- Add test to validate core data migration from model 60 to 61.

### Testing
- Run the app and make sure that Orders are loading properly.
- Replace a build from `trunk` with a build from this branch. Observer that there are no core data errors or warnings and the migration is smooth.

### Screenshots
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
OrderTaxLine in core-data
- Increment core-data model version and add `OrderTaxLine`.
- Configure relationships with `Order` and `OrderItemAttribute`.
- Add `taxes` property in `Order` to hold `OrderTaxLine`.

Commit:
Update `MIGRATIONS.md` about adding Model 61 for  `OrderTaxLine`.

Commit:
Add `loadOrderTaxLine` to `StorageType` to load `OrderTaxLine`.

Commit:
Add test to validate `loadOrderTaxLine` method.

Commit:
Add test to validate core data model 60 to 61.

Commit:
Merge branch 'issue/5809-tax-lines-in-networking' into issue/5809-tax-lines-in-storage

Commit:
Merge branch 'issue/5809-tax-lines-in-networking' into issue/5809-tax-lines-in-storage

Commit:
Update comment with `taxes` to match property name.

Commit:
Merge branch 'trunk' into issue/5809-tax-lines-in-storage

itsmeichigo commented on Storage/Storage/Model/MIGRATIONS.md:
```diff
@@ -2,6 +2,11 @@
+## Model 61 (Release 8.4.0.0)
+- @selanthiraiyan 2022-01-13
+    - Add `OrderTaxLine` entity.
+    - Add `taxes` relationship from `Order` to `OrderTaxLine`.
```
Super nit: Typo "Added" 😅

selanthiraiyan replied:
Fixed in - https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/a970e1881e946c5bfffc13e7c64217c73f60b3a1

itsmeichigo commented on Storage/Storage/Model/WooCommerce.xcdatamodeld/Model 61.xcdatamodel/contents:
```diff
@@ -0,0 +1,786 @@
+        <attribute name="ratePercent" optional="YES" attributeType="Double" defaultValueString="0.0" usesScalarValueType="YES"/>
+        <attribute name="taxID" attributeType="Integer 64" defaultValueString="0.0" usesScalarValueType="NO"/>
+        <attribute name="totalShippingTax" optional="YES" attributeType="String"/>
+        <attribute name="totalTax" optional="YES" attributeType="String"/>
```
For attributes of primitive types, we usually keep them non-optional and use appropriate default values for simplicity. In fact, I think you're keeping them non-optional on the `OrderTaxLine+CoreDataProperties` file, so I think we should update them on the model too.

selanthiraiyan replied:
I understand. Made the necessary changes in the following commits,

`ratePercent` - https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/68e1a1241bfa857c2d0d0ad6175f7334615bd496
`rateID` - https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/089fb576c5315b7cda627e08b8ab0c56f8aaa508
`isCompoundTaxRate` - https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/1e2c8608b2a134fca4fcfb0fa5b4b40ef06914ab

selanthiraiyan replied:
> I left a few comments below, please check it out

I have addressed the comments and PR is ready for another look. Thank you!

itsmeichigo commented on Storage/StorageTests/CoreData/MigrationTests.swift:
```diff
@@ -892,6 +892,40 @@ final class MigrationTests: XCTestCase {
+        let sourceContext = sourceContainer.viewContext
+
+        insertOrder(to: sourceContext)
+        try sourceContext.save()
```
I think it's worth it to double-check that the `taxes` attribute is not present in the old version but is available in the new version. WDYT?

selanthiraiyan replied:
I agree. Added the check in - https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/115dc64f59a523212873e7e6cdd1443d3010b464

itsmeichigo commented on Storage/Storage/Model/OrderTaxLine+CoreDataProperties.swift:
```diff
@@ -0,0 +1,43 @@
+
+}
+
+extension OrderTaxLine: Identifiable {
```
❓ I may be missing something here, but why do we need this conformance here?

selanthiraiyan replied:
This was auto-generated by Xcode. Removed in https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/e0be8a9f71439019bcb866c2ca9ff45792431ffa

itsmeichigo reviewed (commented):
✅ The app builds and runs successfully with this update.

I left a few comments below, please check it out 😄

Commit:
Merge branch 'trunk' into issue/5809-tax-lines-in-storage

Commit:
Fix typo - Use "Added" in migrations document.

Commit:
OrderTaxLine - Remove auto generated `Identifiable` conformance.

Commit:
OrderTaxLine in Model 61
- Make `ratePercent` non optional.

Commit:
OrderTaxLine in Model 61
- Make `rateID` non optional.

Commit:
`isCompoundTaxRate ` in `OrderTaxLine` in Model 61
- Make `isCompoundTaxRate` non optional.
- Use "NO" as default value for `isCompoundTaxRate`.

Commit:
Test that `taxes` relationship appears only after migration.

selanthiraiyan commented:
> I left a few comments below, please check it out

I have addressed your comments by making the necessary changes. And the PR is ready for another look. Thank you!

itsmeichigo commented on Storage/Storage/Model/WooCommerce.xcdatamodeld/Model 61.xcdatamodel/contents:
```diff
@@ -0,0 +1,786 @@
+        <attribute name="rateCode" optional="YES" attributeType="String"/>
+        <attribute name="rateID" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="ratePercent" attributeType="Double" defaultValueString="0.0" usesScalarValueType="YES"/>
+        <attribute name="taxID" attributeType="Integer 64" defaultValueString="0.0" usesScalarValueType="NO"/>
```
I wonder if we should use scalar type for this attribute too 🤔

selanthiraiyan replied:
Fixed in - https://github.com/woocommerce/woocommerce-ios/pull/5918/commits/f939dfbcd59b095ba48c020ce645b33098ec93c3

itsmeichigo reviewed (commented):
All look good! Except for one point - I think we should default to using scalar type for all attributes of primitive types. WDYT?

Commit:
OrderTaxLine core data entity
- Make `taxID` use scalar type.

itsmeichigo reviewed (approved):
🚀
